### PR TITLE
feat: initial federated learning use case implementation

### PR DESCRIPTION
**Pull request by ferrarimarco**
This PR introduces the Federated Learning core platform use case.

As a first step to get a feeling of how it is to integrate and existing use case with the platform, we provision only a simple resource (an Artifact Registry repository).

### CODE REVIEW

**Comment by arueth** [comment]
Once your initial PR #48 is merged, you will want to create an `int-federated-learning` branch. The `int-`, for integration, prefix adds protections to the branch such as requiring a PR to commit code. Then you can create your own personal branch off of that branch to work on.

**Comment by ferrarimarco** [comment]
> Once your initial PR #48 is merged, you will want to create an `int-federated-learning` branch. The `int-`, for integration, prefix adds protections to the branch such as requiring a PR to commit code. Then you can create your own personal branch off of that branch to work on.

Done.

**Comment by ferrarimarco** [comment]
@arueth Can you please have an early look at this? Thanks a lot

**Comment by ferrarimarco** [comment]
Also, I suppose we need two additional build triggers (`uc-federated-learning-terraform-destroy` and `uc-federated-learning-terraform`), right?

**Review by arueth — COMMENTED** [review]
We can sync up on how we can structure the use-case directory to make use of the share_config.

File: platforms/gke/base/use-cases/federated-learning/terraform/initialize/versions.tf [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      version = "6.12.0"
+    }
+    local = {
+      source  = "hashicorp/local"
+      version = "2.5.2"
+    }
+  }
+
+  provider_meta "google" {
+    module_name = "cloud-solutions/acp_gke_base_core_initialize_deploy-v1"
```
[the remark this answers is not in the document]
Comment (ferrarimarco) [reply platforms/gke/base/use-cases/federated-learning/terraform/initialize/versions.tf:30]: right. copy-paste mistake. Fixed.

File: test/ci-cd/cloudbuild/uc-federated-learning-terraform.yaml [hunk 2]
```diff
@@ -36,8 +36,33 @@ steps:
...
+        export ACP_PLATFORM_BASE_DIR="/workspace/platforms/gke/base"
+        export ACP_PLATFORM_CORE_DIR="/workspace/platforms/gke/base/core"
+
+        export TF_VAR_cluster_project_id="${PROJECT_ID}"
+        export TF_VAR_platform_name="commit-${SHORT_SHA}-fl"
+        export TF_VAR_terraform_project_id="${PROJECT_ID}"
+
+        # Deploy the base platform
+        platforms/gke/base/core/deploy.sh
+        platforms/gke/base/core/deploy.sh
```
[the remark this answers is not in the document]
Comment (ferrarimarco) [reply test/ci-cd/cloudbuild/uc-federated-learning-terraform.yaml:54]: Got it. Removed.

File: test/ci-cd/cloudbuild/uc-federated-learning-terraform-destroy.yaml [hunk 3]
```diff
@@ -36,12 +36,24 @@ steps:
...
+        export TERM="xterm"
+        export TF_CLI_ARGS="-no-color"
+        export TF_IN_AUTOMATION="1"
+
+        export ACP_REPO_DIR="/workspace"
+        export ACP_PLATFORM_BASE_DIR="/workspace/platforms/gke/base"
+        export ACP_PLATFORM_CORE_DIR="/workspace/platforms/gke/base/core"
+
+        export TF_VAR_cluster_project_id="${PROJECT_ID}"
+        export TF_VAR_platform_name="commit-${SHORT_SHA}-fl"
```
[the remark this answers is not in the document]
Comment (ferrarimarco) [reply test/ci-cd/cloudbuild/uc-federated-learning-terraform-destroy.yaml:49]: makes sense. Done.

File: platforms/gke/base/use-cases/federated-learning/terraform/container_image_repository/versions.tf [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+  required_providers {
+    google = {
+      source  = "hashicorp/google"
+      version = "6.12.0"
+    }
+  }
+
+  provider_meta "google" {
+    module_name = "cloud-solutions/acp_gke_federated_learning_container_image_repository_deploy-v1"
```
[the remark this answers is not in the document]
Comment (ferrarimarco) [reply platforms/gke/base/use-cases/federated-learning/terraform/container_image_repository/versions.tf:26]: Shortened just to be on the safe side.

File: platforms/gke/base/_shared_config/uc_federated_learning.auto.tfvars [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2024 Google LLC
```
[the remark this answers is not in the document]
Comment (ferrarimarco) [reply platforms/gke/base/_shared_config/uc_federated_learning.auto.tfvars:1]: Done.

**Pull request merged**